docs/JAVA_GRAMMAR_UPDATES.md

```diff
@@ -0,0 +1,228 @@
+  complexity of JavaAstVisitor to ease maintenance.
+
+### Update JavaAstVisitor
+
+Now, our lexer recognizes the `when` keyword, and our parser recognizes the `when` expression. We
+have updated
+TokenTypes to include `LITERAL_WHEN`. At this point, we are able to parse the new syntax, however,
+the new tokens
+will not appear in the AST unless we update the JavaAstVisitor. This is because our ANTLR grammar
+provides us with a parse tree, which we use the JavaAstVisitor to traverse and build our AST.
+
+```java
+
+@Override
+public DetailAstImpl visitGuardedPattern(JavaLanguageParser.GuardedPatternContext ctx) {
+    // since the `guard` rule is a terminal rule, we need to create a new AST node for it
+    final DetailAstImpl guardAstNode = flattenedTree(ctx.guard());
+    // we add the children of the `primaryPattern` and `expression` rules to the AST node
+    guardAstNode.addChild(visit(ctx.primaryPattern()));
+    guardAstNode.addChild(visit(ctx.expression()));
+    return guardAstNode;
+}
+```
+
+Above is an example of the transformation of the `guardedPattern` rule in the JavaAstVisitor. We
+create a new AST node for the `guard` rule, and add the children of the `primaryPattern` and
+`expression` rules to the AST node. This is a simplified example, and the actual implementation
+may vary depending on the complexity of the rule.
+
+Notes:
+ - When we visit a parser rule context, we have access to all of the parse tree elements that
+   correspond to that rule. We can use this information to build our AST.
+ - The structure of the AST should be carefully considered; we want to ease static analysis by
+   providing a clear and concise representation of the source code. We can review existing AST 
+   nodes within other contexts to get an idea of how to structure the AST; however, there is no
+   strict rules regarding this topic. This part of the process relies heavily on heuristics and
+   experience.
+
+### Update Tests
+
+Finally, we need to update our tests to ensure that the new language feature is correctly parsed and
+analyzed. You
+can find our AST
+tests [here](https://github.com/checkstyle/checkstyle/tree/master/src/test/java/com/puppycrawl/tools/checkstyle/grammar).
+
+Notes:
+ - We typically separate tests into directories based on the language level.
+ - We test the content of the AST only, we generally do not test the actual parse tree.
+
+## Additional Considerations
+
+### DetailAstPair
+
+The `DetailAstPair` class is used to represent a pair of AST nodes. It is used in
+the `JavaAstVisitor` to
+represent the nested parent-child relationship between AST nodes, especially where there is
+recursive nesting of only two to three types of tokens. An example of this is the `qualifiedName`
+rule, where the parent node is a `DOT` token and the child nodes are `IDENT` tokens. We use
+`DetailAstPair` to represent this relationship in the JavaAstVisitor#visitQualifiedName method.
+
+```java
+
+### Imaginary Tokens
```

Comment:
You probably forgot the code of `JavaAstVisitor#visitQualifiedName` or unnecessary code block formatting (```Java) should be removed

Reply:
Yep, deleted